File: train/benchmark.ts
// Head-to-head benchmark between two policies. Examples:
//   npm run benchmark -- --a train/checkpoints/latest.json --b random
//   npm run benchmark -- --a train/checkpoints/ckpt-200000.json --b pip --games 2000
// Training-only: nothing under src/ may import from train/.
import { readFileSync } from 'node:fs';
import { parseArgs } from 'node:util';
import { greedyPipAgent, MoveChooser, neuralAgent, randomAgent } from '../src/ai/agents';
import { deserializeNet } from '../src/ai/network';
import { mulberry32 } from '../src/engine/rng';
import { playSeries } from './selfplay';

const { values: args } = parseArgs({
  options: {
    a: { type: 'string', default: 'train/checkpoints/latest.json' },
    b: { type: 'string', default: 'random' },
    games: { type: 'string', default: '1000' },
    seed: { type: 'string', default: '7' },
  },
});

function chooserFor(spec: string): { chooser: MoveChooser; label: string } {
  if (spec === 'random') return { chooser: randomAgent, label: 'random' };
  if (spec === 'pip') return { chooser: greedyPipAgent, label: 'greedy-pip' };
  const { net, meta } = deserializeNet(readFileSync(spec, 'utf8'));
  return { chooser: neuralAgent(net), label: `${spec} (${meta.games} games)` };
}

const a = chooserFor(args.a!);
const b = chooserFor(args.b!);
const games = Number(args.games);

console.log(`A: ${a.label}\nB: ${b.label}\n${games} games, alternating colors…`);
const r = playSeries(a.chooser, b.chooser, games, mulberry32(Number(args.seed)));

const decided = r.aWins + r.bWins;
const winRate = r.aWins / decided;
// Binomial standard error on the win rate.
const se = Math.sqrt((winRate * (1 - winRate)) / decided);
const ppg = (r.aPoints - r.bPoints) / decided;

console.log(
  `\nA wins ${r.aWins}/${decided} = ${(100 * winRate).toFixed(1)}% ± ${(100 * se).toFixed(1)}%`,
);
console.log(`A points-per-game (cubeless): ${ppg >= 0 ? '+' : ''}${ppg.toFixed(3)}`);
console.log(
  `gammons  A ${r.aGammons}, B ${r.bGammons} | backgammons  A ${r.aBackgammons}, B ${r.bBackgammons}`,
);
console.log(
  `avg half-moves ${(r.halfMoves / r.games).toFixed(1)}` +
    (r.aborted ? ` | ABORTED ${r.aborted}` : ''),
);

File: python/bgtrain/benchmark.py
"""Benchmark a trained net vs baselines or another checkpoint.

    uv run python -m bgtrain.benchmark --a checkpoints/latest.json --b pip --games 1000
"""

from __future__ import annotations

import argparse
import math

import torch

from .agents import neural_chooser, pip_chooser, play_match, random_chooser
from .net import load_net


def _chooser(spec, device, gen):
    if spec == "random":
        return random_chooser(gen), "random"
    if spec == "pip":
        return pip_chooser(), "greedy-pip"
    net = load_net(spec, device=device)
    return neural_chooser(net), spec


def main():
    p = argparse.ArgumentParser()
    p.add_argument("--a", default="checkpoints/latest.json")
    p.add_argument("--b", default="random")
    p.add_argument("--games", type=int, default=1000)
    p.add_argument("--seed", type=int, default=7)
    p.add_argument("--device", default="cpu")
    args = p.parse_args()

    gen = torch.Generator(device=args.device).manual_seed(args.seed)
    a, la = _chooser(args.a, args.device, gen)
    b, lb = _chooser(args.b, args.device, gen)
    print(f"A: {la}\nB: {lb}\n{args.games} games, alternating colors…")

    r = play_match(a, b, args.games, args.device, gen)
    decided = max(1, r["decided"])
    wr = r["a_wins"] / decided
    se = math.sqrt(wr * (1 - wr) / decided)
    ppg = (r["a_points"] - r["b_points"]) / r["games"]
    print(f"\nA wins {r['a_wins']}/{decided} = {100 * wr:.1f}% ± {100 * se:.1f}%")
    print(f"A points-per-game (cubeless): {ppg:+.3f}")


if __name__ == "__main__":
    main()
